src/AddButton.js

```diff
@@ -0,0 +1,150 @@
+  let blockDefinitions;
+  if (parentId === null) {
+    blockDefinitions = state[fieldId].blockDefinitions;
+  } else {
+    blockDefinitions = getNestedBlockDefinition(state, fieldId,
+                                                parentId).children;
+  }
+
+  let index = 0;
+  if (blockId !== undefined) {
+    // Incremented by 1 to add after the current block.
+    index = getSiblingsIds(state, fieldId, blockId).indexOf(blockId) + 1;
+  }
+
+  return {blockDefinitions, index};
+}, (dispatch, props) => {
+  const {fieldId, parentId} = props;
+  return bindActionCreators({
+    addBlock: (index, blockType) => addBlock(fieldId, parentId,
+                                             index, blockType),
+  }, dispatch);
+})
+class AddButton extends React.Component {
+  static propTypes = {
+    fieldId: PropTypes.string.isRequired,
+    parentId: PropTypes.string,
+    blockId: PropTypes.string,
+    open: PropTypes.bool,
+    visible: PropTypes.bool,
+  };
+
+  static defaultProps = {
+    open: false,
+    visible: true,
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {open: props.open};
+  }
+
+  get hasChoice() {
+    return this.props.blockDefinitions.length !== 1;
+  }
+
+  toggle = event => {
+    event.preventDefault();
+    event.stopPropagation();
+    if (this.hasChoice) {
+      this.setState({open: !this.state.open});
+    } else {
+      this.props.addBlock(this.props.index,
+                          this.props.blockDefinitions[0].key);
+    }
+  };
+
+  addHandler = event => {
+    event.preventDefault();
+    event.stopPropagation();
+    this.props.addBlock(this.props.index,
+                        event.target.closest('button').value);
+    this.toggle(event);
+  };
+
+  getIcon(blockDefinition) {
+    const {icon} = blockDefinition;
+    if (isNA(icon)) {
+      return null;
+    }
+    return <span className="icon" dangerouslySetInnerHTML={
+      {__html: icon}
+    } />;
+  }
+
+  get panelHeight() {
+    return this.state.open && this.props.visible ? 'auto' : 0;
+  }
+
+  get groupedBlockDefinitions() {
+    const grouped = {};
+    for (const blockDefinition of this.props.blockDefinitions) {
+      let key = blockDefinition.group;
+      if (key === undefined) {
+        key = '';
+      }
+      let others = grouped[key];
+      if (others === undefined) {
+        others = [];
+      }
+      others.push(blockDefinition);
+      grouped[key] = others;
+    }
+    return grouped;
+  }
+
+  render() {
+    const {visible} = this.props;
+    const button = (
+      <button onClick={this.toggle}
+              className={classNames(
+                'add', visible && 'visible',
+                (this.state.open && this.hasChoice) && 'close')}>
+        <i>+</i>
+      </button>
+    );
+    if (this.hasChoice) {
+      return (
+        <React.Fragment>
+          {button}
+          <AnimateHeight height={this.panelHeight} easing="ease-in-out"
+                         contentClassName="add-panel">
```

Comment:
Unfortunately, there is absolutely no way around this. If we want collapse transitions, we have to do this. This is an extremely annoying limitation of CSS3, you can transition between `height: 0` and `height: auto`.

You should not worry about the amount of DOM nodes introduced by `AnimateHeight`. It consists only of 2 nested divs, and every time I use `AnimateHeight` I use the child `div` as a container that I would have done otherwise, by specifying `contentClassName`.